modules/distributor/forwarder/manager.go

```diff
@@ -0,0 +1,370 @@
+
+	for {
+		select {
+		case <-ctx.Done():
+			if err := m.shutdown(); err != nil {
+				return fmt.Errorf("failed to shutdown: %w", err)
+			}
+
+			return nil
+		case <-ticker.C:
+			m.handleTick()
+		}
+	}
+}
+
+func (m *Manager) stop(err error) error {
+	if err != nil {
+		_ = level.Warn(m.logger).Log("msg", "manager returned error from running state", "err", err)
+	}
+
+	return nil
+}
+
+func (m *Manager) handleTick() {
+	tenantToForwarderNames := m.readOverrides()
+	m.updateTenantToForwarderList(tenantToForwarderNames)
+}
+
+// readOverrides returns a mapping between tenant ID and a
+// list of requested forwarders.
+func (m *Manager) readOverrides() map[string][]string {
+	result := make(map[string][]string)
+
+	for _, tenantID := range m.overrides.TenantIDs() {
+		names := m.overrides.Forwarders(tenantID)
+		if len(names) < 1 {
+			continue
+		}
+
+		result[tenantID] = names
+	}
+
+	return result
+}
+
+func (m *Manager) updateTenantToForwarderList(tenantToForwarderNames map[string][]string) {
+	m.tenantToQueueListMu.Lock()
+	defer m.tenantToQueueListMu.Unlock()
+
+	oldTenantIDs := make([]string, 0, len(m.tenantToQueueList))
+	for oldTenantID := range m.tenantToQueueList {
+		oldTenantIDs = append(oldTenantIDs, oldTenantID)
+	}
+
+	newTenantIDs := make([]string, 0, len(tenantToForwarderNames))
+	for newTenantID := range tenantToForwarderNames {
+		newTenantIDs = append(newTenantIDs, newTenantID)
+	}
+
+	diff := diff(oldTenantIDs, newTenantIDs)
+
+	for _, removedTenantID := range diff.removed {
+		m.removeQueueListForTenantUnderLock(removedTenantID)
+	}
+
+	for _, addedTenantID := range diff.added {
+		m.updateQueueListForTenantUnderLock(addedTenantID, tenantToForwarderNames[addedTenantID])
+	}
+
+	// Go through all tenants and try to update them.
+	// If forwarder names for particular tenant changed, this will update them.
+	// If names didn't change, this is a noop.
+	for tenantID := range m.tenantToQueueList {
+		m.updateQueueListForTenantUnderLock(tenantID, tenantToForwarderNames[tenantID])
+	}
+}
+
+func (m *Manager) removeQueueListForTenantUnderLock(tenantID string) {
+	ql, found := m.tenantToQueueList[tenantID]
+	if !found {
+		_ = level.Warn(m.logger).Log("msg", "queue list not found", "tenantID", tenantID)
+		return
+	}
+
+	if err := ql.shutdown(context.TODO()); err != nil {
+		_ = level.Warn(m.logger).Log("msg", "failed to shutdown queue list", "tenantID", tenantID)
+	}
+
+	delete(m.tenantToQueueList, tenantID)
+}
+
+// updateQueueListForTenantUnderLock should only be called under tenantToQueueListMu write lock.
+func (m *Manager) updateQueueListForTenantUnderLock(tenantID string, forwarderNames []string) {
+	queueList, found := m.tenantToQueueList[tenantID]
+	if !found {
+		queueList = newQueueList(m.logger, m.reg, tenantID)
+		m.tenantToQueueList[tenantID] = queueList
+	}
+
+	queueList.update(forwarderNames, m.forwarderNameToForwarder)
+}
+
+func (m *Manager) getQueueList(tenantID string) (*queueList, bool) {
```

Comment:
if this method is only called from `ForTenant()` then why don't we remove it and place the contents in `ForTenant()`?

Reply:
I see no value in removing this function

Reply:
less code?

Reply:
Done in [1871143](https://github.com/grafana/tempo/pull/1775/commits/187114373c152431a9ecdda8f4cf9136d3e81cca).